File: src/main/java/cn/comicalpixel/creeperstarbedwars/Items/FIREBALL_Item.java
package cn.comicalpixel.creeperstarbedwars.Items;

import cn.comicalpixel.creeperstarbedwars.Arena.GamePlayers;
import cn.comicalpixel.creeperstarbedwars.Arena.Stats.GameStats;
import cn.comicalpixel.creeperstarbedwars.Config.ConfigData;
import cn.comicalpixel.creeperstarbedwars.Listener.PlayerBlocks;
import cn.comicalpixel.creeperstarbedwars.Utils.FireballUtil;
import cn.comicalpixel.creeperstarbedwars.Utils.LocationUtil;
import cn.comicalpixel.creeperstarbedwars.Utils.MessageVariableUtils;
import org.bukkit.*;
import org.bukkit.block.Block;
import org.bukkit.entity.Fireball;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.block.Action;
import org.bukkit.event.entity.EntityDamageByEntityEvent;
import org.bukkit.event.entity.EntityDamageEvent;
import org.bukkit.event.entity.EntityExplodeEvent;
import org.bukkit.event.player.PlayerInteractEvent;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.ItemMeta;
import org.bukkit.util.Vector;

import java.util.*;

public class FIREBALL_Item implements Listener {

    private Map<Player, Long> cooldownMap = new HashMap<>();

    @EventHandler
    public void PlayerInteract(PlayerInteractEvent e) {

        if (GameStats.get() != 2 && GameStats.get() != 3) {return;}

        Player p = e.getPlayer();

        if (e.getItem() == null) return;
        if (e.getItem().getType() != Material.FIREBALL) return;
        if (e.getAction() != Action.RIGHT_CLICK_AIR && e.getAction() != Action.RIGHT_CLICK_BLOCK) return;

        e.setCancelled(true);

        if (cooldownMap.containsKey(p) && System.currentTimeMillis() - cooldownMap.get(p) < (ConfigData.ItemsInGame_fireball_cooldown * 1000L)) {
            int cooldms = (ConfigData.ItemsInGame_fireball_cooldown * 1000) - (int) (System.currentTimeMillis() - cooldownMap.get(p));
            int cooldms_s = MessageVariableUtils.convertMillisecondsToSeconds(cooldms);
            String cooldmess = String.valueOf(cooldms_s); // String.valueOf(cooldownMap.containsKey(player)); //System.currentTimeMillis() -
            e.setCancelled(true);
            p.sendMessage(ConfigData.ItemsInGame_fireball_cooldown_chat.replace("{cooldown}", cooldmess));
            return;
        }

        cooldownMap.put(p, System.currentTimeMillis());

        ItemStack itemInHand = p.getItemInHand();
        if (itemInHand.getAmount() > 1) {
            itemInHand.setAmount(itemInHand.getAmount() - 1);
        } else {
            p.getInventory().remove(itemInHand);
        }


        Fireball f = (Fireball)p.launchProjectile(Fireball.class);
        Vector direction = p.getEyeLocation().getDirection();
        FireballUtil.setFireballDirection(f, direction);
        f.setVelocity(f.getDirection().multiply(5));
        f.setIsIncendiary(false);
        f.setShooter(p);
        f.setYield(ConfigData.ItemsInGame_fireball_radius);

    }


    @EventHandler
    public void EntityExplode(EntityExplodeEvent e) {

        if (GameStats.get() != 2 && GameStats.get() != 3) return;
        if (e.getEntity() instanceof Fireball) {
            e.setCancelled(true);


            // 防爆玻璃 不好的算法:(
            Location explosionLoc = e.getLocation();
            List<Block> blocksToRemove = new ArrayList<>();
//            for (Block db : e.blockList()) {
//
//                boolean blocked = false;
//                Vector direction = db.getLocation().toVector().subtract(explosionLoc.toVector()).normalize();
//
//
//                for (double i = 0; i < db.getLocation().add(0.5, 0, 0.5).distance(explosionLoc.clone().add(0.5, 0, 0.5)); i += 0.2) {
//                    Location checkLoc = explosionLoc.clone().add(direction.clone().multiply(i));
//                    Block block = checkLoc.getBlock();
//                    if (block.getType() == Material.GLASS || block.getType() == Material.STAINED_GLASS) {
//                        blocked = true;
//                        break;
//                    }
//                }
//
//
//                if (!blocked && PlayerBlocks.player_blocks.contains(db) &&
//                        db.getType() != Material.GLASS && db.getType() != Material.STAINED_GLASS) {
//                    blocksToRemove.add(db);
//                }
//            }
            for (Block db : e.blockList()) {

                boolean blocked = false;
                Location blockLoc = db.getLocation().clone().add(0.5, 0.5, 0.5);
                Vector direction = blockLoc.toVector().subtract(explosionLoc.toVector()).normalize();

                for (double i = 0; i < blockLoc.distance(explosionLoc); i += 0.2) {
                    Location checkLoc = explosionLoc.clone().add(direction.clone().multiply(i));

                    for (double x = -0.5; x <= 0.5; x += 0.5) {
                        for (double y = -0.5; y <= 0.5; y += 0.5) {
                            for (double z = -0.5; z <= 0.5; z += 0.5) {
                                Block block = checkLoc.clone().add(x, y, z).getBlock();
                                if (block.getType() == Material.GLASS || block.getType() == Material.STAINED_GLASS) {
                                    blocked = true;
                                    break;
                                }
                            }
                            if (blocked) break;
                        }
                        if (blocked) break;
                    }
                    if (blocked) break;
                }

                if (!blocked && PlayerBlocks.player_blocks.contains(db) &&
                        db.getType() != Material.GLASS && db.getType() != Material.STAINED_GLASS) {
                    blocksToRemove.add(db);
                }
            }


            // 防炸地图方块
//            for (Block db : e.blockList()) {
            for (Block db : blocksToRemove) {
                if (PlayerBlocks.player_blocks.contains(db) &&
                        db.getType() != Material.GLASS && db.getType() != Material.STAINED_GLASS
                ) {
                    db.setType(Material.AIR);
                }
            }


            // KB & DAMAGE
            for (Player p : GamePlayers.players) {
                if (p.getLocation().distance(e.getLocation()) < ConfigData.ItemsInGame_fireball_radius && Objects.equals(p.getLocation().getWorld().getName(), e.getLocation().getWorld().getName())) {
                    if (GameStats.get() == 2) {
                        p.damage(ConfigData.ItemsInGame_fireball_damage);
                    }
                    p.setLastDamageCause(new EntityDamageEvent(p, EntityDamageEvent.DamageCause.ENTITY_EXPLOSION, ConfigData.ItemsInGame_fireball_damage));
                    p.setVelocity(LocationUtil.getPosition(p.getLocation(), e.getLocation(), 0.9).multiply(ConfigData.ItemsInGame_fireball_vlocity_multiply).setY(ConfigData.ItemsInGame_fireball_vlocity_y));
                }
            }



            // 播放粒子效果
            for (Player p : Bukkit.getOnlinePlayers()) {
                p.playEffect(e.getLocation(), Effect.EXPLOSION_LARGE, 2);
            }


        }

    }

    @EventHandler
    public void onPlayerDamageforEntity(EntityDamageByEntityEvent e) {
        if (GameStats.get() != 2 && GameStats.get() != 3) return;
        if (e.getDamager() instanceof Fireball) {
            e.setCancelled(true);
        }
    }


}

File: src/main/kotlin/cn/comicalpixel/creeperstarbedwars/Items/FIREBALL_Item.kt
package cn.comicalpixel.creeperstarbedwars.Items

import cn.comicalpixel.creeperstarbedwars.Arena.Stats.GameStats
import cn.comicalpixel.creeperstarbedwars.Config.ConfigData
import cn.comicalpixel.creeperstarbedwars.Listener.PlayerBlocks
import cn.comicalpixel.creeperstarbedwars.Utils.FireballUtil
import cn.comicalpixel.creeperstarbedwars.Utils.LocationUtil
import org.bukkit.Bukkit
import org.bukkit.ChatColor
import org.bukkit.Effect
import org.bukkit.Material
import org.bukkit.entity.Fireball
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.block.Action
import org.bukkit.event.entity.EntityDamageByEntityEvent
import org.bukkit.event.entity.EntityDamageEvent
import org.bukkit.event.entity.EntityExplodeEvent
import org.bukkit.event.player.PlayerInteractEvent

class FIREBALL_Item : Listener {

    private val cooldownMap: MutableMap<Player, Long> = HashMap()

    @EventHandler
    fun onPlayerInteract(e: PlayerInteractEvent) {
        val p = e.player

        if (e.item != null) {
            if (e.item.type == Material.FIREBALL) {
                if (e.action == Action.RIGHT_CLICK_AIR || e.action == Action.RIGHT_CLICK_BLOCK) {

                    if (GameStats.get() != 2 && GameStats.get() != 3) return

                    e.isCancelled = true

                    // 检查冷却时间
                    if (cooldownMap.containsKey(p) && System.currentTimeMillis() - cooldownMap[p]!! < (ConfigData.ItemsInGame_fireball_cooldown * 1000)) {
                        val cooldms = (ConfigData.ItemsInGame_fireball_cooldown * 1000) - (System.currentTimeMillis() - cooldownMap[p]!!).toInt()
                        val cooldms_s = convertMillisecondsToSeconds(cooldms.toLong())
                        val cooldmess =
                            cooldms_s.toString() // String.valueOf(cooldownMap.containsKey(player)); //System.currentTimeMillis() -
                        e.isCancelled = true
                        p.sendMessage(ConfigData.ItemsInGame_fireball_cooldown_chat.replace("{cooldown}", cooldmess))
                        return
                    }

                    cooldownMap[p] = System.currentTimeMillis() // 设置当前玩家的冷却时间

                    val itemInHand = p.itemInHand
                    if (itemInHand.amount > 1) {
                        itemInHand.amount = itemInHand.amount - 1
                    } else {
                        p.inventory.remove(itemInHand)
                    }

                    //                    Fireball fireball = (Fireball) p.launchProjectile(Fireball.class); //.getLocation().add(0,0.5,0);
//                    // Fireball fireball = p.getWorld().spawn(p.getLocation().add(0,1,0), Fireball.class);
//                    fireball.setBounce(false); // 设置火焰弹不反弹
//                    fireball.setShooter(p);
//                    fireball.setYield(1); // 设置火焰弹的爆炸威力
//                    Vector direction = p.getLocation().getDirection().multiply(0.5);
//                    fireball.setVelocity(direction);// 设置发射者为玩家
                    val f = p.launchProjectile(Fireball::class.java) as Fireball
                    val direction = p.eyeLocation.direction
                    FireballUtil.setFireballDirection(f, direction)
                    f.velocity = f.direction.multiply(5)
                    f.setIsIncendiary(false)
                    f.shooter = p
                    f.yield = 3f

                    // 获取玩家的视线方向
                    // Vector direction = p.getLocation().getDirection();
                    //Vector direction = p.getLocation().getDirection().multiply(0.1);

// 设置火焰弹的飞行方向
                    // fireball.setDirection(direction);
                }
            }
        }
    }

    @EventHandler
    fun onEntityExplode(e: EntityExplodeEvent) {
        if (e.entity is Fireball) {
            // 取消事件，防止生成火焰和破坏方块
            e.isCancelled = true

            // 获取爆炸位置
            val location = e.location
            // 获取世界
            val world = location.world

            for (b in e.blockList()) {
                if (PlayerBlocks.player_blocks.contains(b) && (b.type != Material.GLASS || b.type != Material.STAINED_GLASS)) {
                    b.type = Material.AIR
                }
            }

            for (p in world.players) {
                if (p.location.distance(e.location) < 3) {
                    if (GameStats.get() == 2) {
                        p.damage(ConfigData.ItemsInGame_fireball_damage - 0.1)
                    }
                    p.lastDamageCause = EntityDamageEvent(p, EntityDamageEvent.DamageCause.ENTITY_EXPLOSION, 3)
                    p.velocity = LocationUtil.getPosition(p.location, e.location, 0.9).multiply(ConfigData.ItemsInGame_fireball_vlocity_multiply)
                }
            }

            for (p in Bukkit.getOnlinePlayers()) {
                p.playEffect(e.location, Effect.EXPLOSION_LARGE, 2)
            }
        }
    }

    @EventHandler
    fun onPlayerDamageforEntity(e: EntityDamageByEntityEvent) {
        if (e.damager is Fireball) {
            e.isCancelled = true
        }
    }


    fun convertMillisecondsToSeconds(milliseconds: Long): Int {
        return (milliseconds / 1000 + 1).toInt() //+1实现防出现0秒
    }

}